### PR TITLE
Cosine Learning rate

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Implements Cosine Learning Rate annealing with warm restarts ("SGDR", https://arxiv.org/abs/1608.03983).

Also adds `--lr-period-updates` which can stretch the length of `--lr-scheduler invsqrt` to be different than that of `--warmup-updates`. The default behavior remains the same still.

Needs to be resolved with #1812 still. I don't think it will work when merged into master.

Additionally, I think this code is begging for a refactor now, but not sure if I want to put in the love right now.

**Testing steps**
Desperately needs a unit test. 

**Logs**
```
If applicable, please paste the command line from your testing:
```

**Other information**
Any other information or context you would like to provide.

**Data tests (if applicable)**
If you added a new teacher, you will be asked to run
`python setup.py test -s tests.suites.datatests`. Please paste this log here.

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -1662,6 +1706,52 @@ def update_params(self):
...
+                passes = np.floor(np.log((lr_mult - 1) * n + 1) / np.log(lr_mult))
+                last_reset_step = (
+                    self._lr_period * (1 - lr_mult ** passes) / (1 - lr_mult)
+                )
+                step = cosine_steps - last_reset_step
+                # cycle_steps contains the total length of this period
+                cycle_steps = self._lr_period * lr_mult ** passes
+                self.scheduler.T_max = cycle_steps
+
+            # the height of the peak of the cosine wave is multiplied by
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: improve this comment

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -1662,6 +1706,52 @@ def update_params(self):
...
+            warmup_updates = max(0, self.opt['warmup_updates'])
+            cosine_steps = self._number_training_updates - warmup_updates
+
+            # factor we lengthen the cosine period by
+            lr_mult = self.opt['lr_period_mult']
+
+            # passes contains the number of periods we have iterated
+            # step contains the number of steps relative to the last reset
+            if lr_mult == 1:
+                # special case, do this
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: improve this comment

File: parlai/core/torch_agent.py [hunk 3]
```diff
@@ -420,88 +420,119 @@ def add_cmdline_args(cls, argparser):
...
             'fixed scheduler, LR is decayed every <patience> validations.',
         )
-        agent.add_argument(
+        lr_group.add_argument(
             '--lr-scheduler-decay',
             type=float,
             default=0.5,
-            help='Decay factor for LR scheduler, or how much LR is multiplied by '
-            'when it is lowered.',
+            help=(
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: do you need these parentheses

File: parlai/core/torch_agent.py [hunk 4]
```diff
@@ -420,88 +420,119 @@ def add_cmdline_args(cls, argparser):
...
+                'Decay factor for LR scheduler, or how much LR is multiplied by '
+                'when it is lowered (on plataue with --lr-scheduler reduceonplateau '
+                'and on periods in --lr-scheduler cosine)'
+            ),
         )
-        agent.add_argument(
+        lr_group.add_argument(
+            '--lr-period-mult',
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: would be nice if we were able to have nested args ... maybe some day!

File: parlai/core/torch_agent.py [hunk 5]
```diff
@@ -844,6 +876,10 @@ def _warmup_lr(step):
         else:
             self.warmup_scheduler = None
 
+        self._lr_period = self.opt.get('lr_period_updates')
+        if self._lr_period is None or self._lr_period <= 0:
+            self._lr_period = self.opt.get('warmup_updates', -1)
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: why do we set it to warmup updates here?

File: parlai/core/torch_agent.py [hunk 6]
```diff
@@ -1662,6 +1706,52 @@ def update_params(self):
...
+                step = cosine_steps - last_reset_step
+                # cycle_steps contains the total length of this period
+                cycle_steps = self._lr_period * lr_mult ** passes
+                self.scheduler.T_max = cycle_steps
+
+            # the height of the peak of the cosine wave is multiplied by
+            # decay every period
+            mult_max = self.opt['lr_scheduler_decay'] ** passes
+            lr = self.opt['learningrate']
+            self.scheduler.base_lrs = [lr * mult_max for _ in self.scheduler.base_lrs]
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: what does this do?

File: parlai/core/torch_agent.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: No, my opinion was that it was just more readable this way. You prefer the other?

File: parlai/core/torch_agent.py (hunk 5 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: This is legacy behavior. Currently `--lr-scheduler` uses `warmup_updates` to set the decay rate. This lets the user have separate those values, but still defaults to the old behavior of depending on `--warmup-updates`. (Previously it was actually not possible to use invsqrt without warmup)

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: It's the way you set the peak learning rate. Every time you iterate through a cosine (half period), the coefficient resets to 1, but the max rate decays. I should scare up an illustration.

Comment (stephenroller) [reply parlai/core/torch_agent.py]: Okay here's how it looks. This corresponds to

`--lr-scheduler-decay 0.75 --lr-period-mult 2 --lr-period-updates 1000`

Screen Shot 2019-07-10 at 10 51 03 AM

Notice at the end of a period, the peak is multiplied by `0.75` and the length of the new period is doubled.

File: parlai/core/torch_agent.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: doesn't really matter, other way looks a bit more consistent

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: nice thanks for the picture

**Comment by emilydinan** [comment]
curious, have you tried this and found that it helps?

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (mwillwork) [reply parlai/core/torch_agent.py]: The paper sometimes uses a schedule/not a multiplicative decay factor (i.e. 0.050, 0.025, 0.01 and 0.005) - do we care?

File: parlai/core/torch_agent.py (hunk 3 above)
Comment (mwillwork) [reply parlai/core/torch_agent.py]: Sorry if stupid/intentional - is there a reason we don't incorporate their LR min value from the paper? (Seems like 0 is always used in the arg to the torch CosineAnnealingLR object). Too much trouble to add another config arg?

Comment (stephenroller) [reply parlai/core/torch_agent.py]: Laziness tbh. I actually added it, it made a bug, and chose to delete the relevant code rather than figure out my bug. It's real unclear to me what's a right value for eta_min.

@glample you use 0? even though you said you only do a single period...

Comment (mwillwork) [reply parlai/core/torch_agent.py]: Well, not a huge deal either way, but 0 seems a bit pointless no? The last few updates (esp the last one) before each restart are a waste in that case

Comment (stephenroller) [reply parlai/core/torch_agent.py]: Only one is a waste, the rest are just REALLY tiny ;)

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: I don’t think so. I’m mimicking fairseq for this bit. Both come out to roughly the same. With decay of 0.5 you get (.5, .25, .125. .0625, ...) so it’s the right ballpark

**Comment by stephenroller** [comment]
@glample told me he likes it with —lr-period-updates 1000000. So you never get to the end of the first period... If we set eta min like @mwillwork suggests we could at least have it be smooth. Getting down to that super tiny LR (1e-8) was helpful for a very final part of the learning, but the sudden jump up hurts.

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (mwillwork) [reply parlai/core/torch_agent.py]: Sure - more a question of if there are times when it wouldn't be similar. Say you might want something other than multiplicative (i.e. a defined schedule or log or something like that). I'm happy to leave it though if you are. I need to get calibrated as to what extent we care about such things!

**Comment by mwillwork** [comment]
@stephenroller - What is our philosophy on implementing exactly what's in a cited paper? i.e. I don't mind if you don't see the need for the min value, but yes if we have a min and set a smart decay factor, the restart need not be so jarring.

**Comment by stephenroller** [comment]
I'm abandoning this because:

1. I have higher priority items, but deserves to be properly experimented with;
2. As Mary points out, it's unfaithful in some ways to the original paper, potentially adding confusion and trouble down the line;
3. Learning rate schedulers deserve a refactor so that their behavior can be somewhat isolated from TorchAgent.

Marking on the TODO & Roadmap for future revisiting.
